MoltenVK/MoltenVK/GPUObjects/MVKDevice.mm

```diff
@@ -3469,6 +3483,27 @@ static uint32_t mvkGetEntryProperty(io_registry_entry_t entry, CFStringRef prope
 	*pPeerMemoryFeatures = VK_PEER_MEMORY_FEATURE_COPY_SRC_BIT | VK_PEER_MEMORY_FEATURE_COPY_DST_BIT;
 }
 
+VkResult MVKDevice::getMemoryHostPointerProperties(VkExternalMemoryHandleTypeFlagBits handleType,
+												   const void* pHostPointer,
+												   VkMemoryHostPointerPropertiesEXT* pMemHostPtrProps) {
+	if (pMemHostPtrProps) {
+		switch (handleType) {
+			case VK_EXTERNAL_MEMORY_HANDLE_TYPE_HOST_ALLOCATION_BIT_EXT:
+				pMemHostPtrProps->memoryTypeBits = _physicalDevice->getHostVisibleMemoryTypes();
+				break;
+			case VK_EXTERNAL_MEMORY_HANDLE_TYPE_HOST_MAPPED_FOREIGN_MEMORY_BIT_EXT:
+				// Vulkan requires only host-visible memory types here, but Metal
+				// only supports Private storage for cross-device buffer sharing.
+				pMemHostPtrProps->memoryTypeBits = 0;
+				break;
```

Comment:
AFAICT, the docs for `-[MTLDevice newBufferWithBytesNoCopy:length:options:deallocator:]` only state that the memory allocation must belong to a single VM region. I don't see how passing the `-[MTLBuffer contents]` of a foreign device's buffer violates that constraint. What happens when you do that?

Also, I don't recall Metal allowing the direct sharing of `MTLBuffer` objects at all. All of the resource-sharing mechanisms I know of (`MTLSharedTextureHandle`, `IOSurface`) only publicly support `MTLTexture`s. (I know about the secret `MTLDevice` methods for creating `MTLBuffer`s from `IOSurface`s, but they're obviously not supported for public use.) The one thing I think you might be thinking of, I believe is intended to be used within a peer group (i.e. a Vulkan device group).

Reply:
> I don't see how passing the `-[MTLBuffer contents]` of a foreign device's buffer violates that constraint.

Ah. Good insight! Thanks.

I agree host-mapped memory allocated on another GPU should be fine to import. I've made changes to support `VK_EXTERNAL_MEMORY_HANDLE_TYPE_HOST_MAPPED_FOREIGN_MEMORY_BIT_EXT `.

>  I don't recall Metal allowing the direct sharing of MTLBuffer objects at all.

macOS Metal supports creating a `MTLBuffer` as a view on a foreign `MTLBuffer`, but only for `Private` storage and BLIT operations, so not useful for this extension:

- [-`[MTBuffer newRemoteBufferViewForDevice:]`](https://developer.apple.com/documentation/metal/mtlbuffer/2967415-makeremotebufferview)
- [-`[MTBuffer remoteStorageBuffer]`](https://developer.apple.com/documentation/metal/mtlbuffer/2967416-remotestoragebuffer)

Reply:
> macOS Metal supports creating a `MTLBuffer` as a view on a foreign `MTLBuffer`, but only for `Private` storage and BLIT operations, so not useful for this extension:
> 
> * [-`[MTBuffer newRemoteBufferViewForDevice:]`](https://developer.apple.com/documentation/metal/mtlbuffer/2967415-makeremotebufferview)
> * [-`[MTBuffer remoteStorageBuffer]`](https://developer.apple.com/documentation/metal/mtlbuffer/2967416-remotestoragebuffer)

Yep, that's the one I thought you were thinking of. The docs also state that this device and the other device must belong to the same peer group. We'd only need it to implement device groups for real.